packages/shopping-cart/src/state-based-actions.ts

```diff
@@ -0,0 +1,68 @@
+import debugFactory from 'debug';
+import { playQueuedActions } from './shopping-cart-reducer';
+import type { ShoppingCartState, ShoppingCartReducerDispatch, CacheStatus } from './types';
+
+const debug = debugFactory( 'shopping-cart:state-based-actions' );
+
+function fetchInitialCart(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { cacheStatus } = state;
+	if ( cacheStatus === 'fresh' && cacheStatus !== lastCacheStatus ) {
+		debug( 'triggering fetch of initial cart' );
+		dispatch( { type: 'FETCH_INITIAL_RESPONSE_CART' } );
+		dispatch( { type: 'GET_CART_FROM_SERVER' } );
+	}
+}
+
+function prepareInvalidCartForSync(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { queuedActions, cacheStatus } = state;
+	if (
+		queuedActions.length === 0 &&
+		cacheStatus === 'invalid' &&
+		cacheStatus !== lastCacheStatus
+	) {
```

Comment:
> the cart will be synced two times:

I don't think that's the case because changes can still be made to an invalid cart and because everything uses the dispatcher to queue actions. Here's a more nuanced timeline of events:

1. addProductsToCart( P1 ) is dispatched and the dispatch is deferred
2. addProductsToCart( P2 ) is dispatched and the dispatch is deferred
3. the P1 change is processed and changes cacheStatus to invalid
4. an action is triggered to set the cacheStatus to pending and the dispatch is deferred
5. an action is triggered to sync the cart with the server and the dispatch is deferred
6. the P2 change is processed so now the local cart has both P1 and P2
7. the cacheStatus change is processed and sets the cart to pending (at this point, any further actions will be queued and would cause a second request later)
8. the sync action is processed and issues a server POST request
9. the server replies, marking cacheStatus as valid
10. this resolves both promises for P1 and P2.

Reply:
Even though I believe the existing system described above works well, and I actually like the relatively elegant logic it uses (everything dispatches, everything is deferred into a queue, things run in the order of the queue), It does clean up the complexity by removing the deferred dispatches, so I made that change.

I still had to defer the state-based actions (and move `notifySubscribers` into there) to allow batching actions, but it all seems to work still. Maybe there's a way to improve it further; I'll wait to see what you think.

Reply:
You're right! I missed that in this step:

> 5. state based actions trigger an action to sync the cart with the server and the dispatch is deferred.

the lock and sync actions are only scheduled rather than performed, so the `P2` update runs before them :+1: